backend/src/apiserver/storage/db.go

```diff
@@ -88,6 +91,11 @@ func (d MySQLDialect) IsDuplicateError(err error) bool {
 	return ok && sqlError.Number == mysqlerr.ER_DUP_ENTRY
 }
 
+// UpdateFromOrJoin TODO(gkcalat): deprecate resource_references table once we migration to v2beta1 is available.
+func (d MySQLDialect) UpdateWithJointOrFrom(targetTable, joinTable, setClause, joinClause, whereClause string) string {
+	return fmt.Sprintf("UPDATE %s JOIN %s ON %s SET %s WHERE %s", targetTable, joinTable, joinClause, setClause, whereClause)
```

Comment:
I just realized that this actually should be `LEFT JOIN` because we are replacing the `OR` op in the `WHERE` clause with this.

Reply:
Yep! Thanks for catching this!

Reply:
Re-opening this conversation... 
I just realized that the below statement is not correct. 
> I just realized that this actually should be `LEFT JOIN` because we are replacing the `OR` op in the `WHERE` clause with this.

This is the original request:
```UPDATE run_details SET StorageState = 'ARCHIVED' WHERE StorageState <> 'ARCHIVED' AND UUID in (SELECT ResourceUUID FROM resource_references as rf WHERE (rf.ReferenceUUID = "d86e5315-09d7-4c8a-982e-4de004b2e4b6" AND rf.ResourceType = 'Run' AND rf.ReferenceType = 'Experiment')) OR ExperimentUUID = 'd86e5315-09d7-4c8a-982e-4de004b2e4b6'; ``` 
created by the following code.
``` 
	// TODO(gkcalat): deprecate resource_references table once we migration to v2beta1 is available.
	// TODO(jingzhang36): use inner join to replace nested query for better performance.
	filteredRunsSql, filteredRunsArgs, err := sq.Select("ResourceUUID").
		From("resource_references as rf").
		Where(sq.And{
			sq.Eq{"rf.ResourceType": model.RunResourceType},
			sq.Eq{"rf.ReferenceUUID": expId},
			sq.Eq{"rf.ReferenceType": model.ExperimentResourceType},
		}).ToSql()
	if err != nil {
		return util.NewInternalServerError(err,
			"Failed to create query to filter the runs in an experiment %s. error: '%v'", expId, err.Error())
	}
	updateRunsSql, updateRunsArgs, err := sq.
		Update("run_details").
		SetMap(sq.Eq{
			"StorageState": model.StorageStateArchived.ToString(),
		}).
		Where(sq.NotEq{"StorageState": model.StorageStateArchived.ToString()}).
		Where(fmt.Sprintf("UUID in (%s) OR ExperimentUUID = '%s'", filteredRunsSql, expId), filteredRunsArgs...).
		ToSql()
	if err != nil {
		return util.NewInternalServerError(err,
			"Failed to create query to archive the runs in an experiment %s. error: '%v'", expId, err.Error())
	}
```

This is the new Optimized MySQL query: 

```UPDATE run_details JOIN resource_references ON run_details.UUID = resource_references.ResourceUUID SET StorageState = "STORAGESTATE_ARCHIVED" WHERE resource_references.ResourceType = "RUN" AND resource_references.ReferenceUUID = "<experiment_id>" AND resource_references.ReferenceType = "Experiment" ```

By adding `LEFT JOIN` we won't directly achieve `OR ExperimentUUID = 'experiment_ID'` from the original query, because:
 1. In general, LEFT JOIN intends to return the entire LEFT table (we just do not want this) and 
 2. The  WHERE clause of the new query is too specific: ` WHERE resource_references.ResourceType = "RUN" AND resource_references.ReferenceUUID = "<experiment_id>" AND resource_references.ReferenceType = "Experiment"`. 
Even if we add an `OR ExperimentUUID = 'experiment_ID'` statement it won't work because of how we JOIN both Tables. Plus there is already a dedicated query that updates all `run_details` based just on `experimentUUID` see (ref A) : https://github.com/kubeflow/pipelines/blob/4f74148343397297bbb89aa7d24cedf7874fb70f/backend/src/apiserver/storage/experiment_store.go#L338-L346 

In short :), After I took a look at how Run and Jobs are created IMO, it would be a bug if we have in the database run_details record without a corresponding `resource_reference` record with `ResourceType= RUN` and `ReferenceType=Experiment`. 
So I suggest keeping the INNER JOIN version of the optimized request and even removing the  (*ref A) - IMO it has been added for "just in case"

Reply:
Thank you @difince for a thorough analysis. I agree.